From 7ccf63d5d88ede0a443923bc8340acd929261dc4 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Mon, 16 Nov 2020 19:36:05 +0900
Subject: [PATCH] udev: save version of udevd which generates the udev database

---
 src/udev/udev-event.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/udev/udev-event.c b/src/udev/udev-event.c
index 64ae83c479..aa9c14b608 100644
--- a/src/udev/udev-event.c
+++ b/src/udev/udev-event.c
@@ -1005,6 +1005,10 @@ int udev_event_execute_rules(UdevEvent *event,
         if (r < 0)
                 log_device_warning_errno(dev, r, "Failed to copy all tags from old database entry, ignoring: %m");
 
+        r = device_add_property(dev, "UDEV_VERSION", STRINGIFY(PROJECT_VERSION));
+        if (r < 0)
+                log_device_warning_errno(dev, r, "Failed to set 'UDEV_VERSION' property, ignoring: %m");
+
         if (sd_device_get_devnum(dev, NULL) >= 0)
                 /* Disable watch during event processing. */
                 (void) udev_watch_end(event->dev_db_clone);
