From 98713edfce87d87f78816ab9180f05ba064063cb Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Mon, 23 Nov 2020 13:28:47 +0900
Subject: [PATCH] network: stop IPv4LL engine when DHCPv4 address is
 successfully acquired

---
 src/network/networkd-dhcp4.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/network/networkd-dhcp4.c b/src/network/networkd-dhcp4.c
index 0f2451cac5..9fb1274067 100644
--- a/src/network/networkd-dhcp4.c
+++ b/src/network/networkd-dhcp4.c
@@ -1105,6 +1105,11 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
                                 return r;
                         }
 
+                        if (link->ipv4ll) {
+                                r = sd_ipv4ll_stop(link->ipv4ll);
+                                if (r < 0)
+                                        log_link_warning_errno(link, r, "Failed to drop IPv4 link-local address, ignoring: %m");
+                        }
                         break;
                 case SD_DHCP_CLIENT_EVENT_RENEW:
                         r = dhcp_lease_renew(client, link);
@@ -1112,6 +1117,12 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
                                 link_enter_failed(link);
                                 return r;
                         }
+
+                        if (link->ipv4ll) {
+                                r = sd_ipv4ll_stop(link->ipv4ll);
+                                if (r < 0)
+                                        log_link_warning_errno(link, r, "Failed to drop IPv4 link-local address, ignoring: %m");
+                        }
                         break;
                 case SD_DHCP_CLIENT_EVENT_IP_ACQUIRE:
                         r = dhcp_lease_acquired(client, link);
@@ -1119,6 +1130,12 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
                                 link_enter_failed(link);
                                 return r;
                         }
+
+                        if (link->ipv4ll) {
+                                r = sd_ipv4ll_stop(link->ipv4ll);
+                                if (r < 0)
+                                        log_link_warning_errno(link, r, "Failed to drop IPv4 link-local address, ignoring: %m");
+                        }
                         break;
                 case SD_DHCP_CLIENT_EVENT_SELECTING:
                         if (!set_isempty(link->network->dhcp_allow_listed_ip)) {
