From a61c0bdf3915495436b36e55e02a1614f6a22cb6 Mon Sep 17 00:00:00 2001
From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Fri, 20 Nov 2020 02:34:48 +0900
Subject: [PATCH] coccinelle: add one more rule to use return value of
 log_xxx_errno()

---
 coccinelle/errno.cocci | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/coccinelle/errno.cocci b/coccinelle/errno.cocci
index ed74c0a98a..c928266481 100644
--- a/coccinelle/errno.cocci
+++ b/coccinelle/errno.cocci
@@ -30,3 +30,12 @@ expression e;
 + return
   log_LEVEL_errno(e, ...);
 - return -e;
+@@
+identifier log_LEVEL_errno =~ "^log_(debug|info|notice|warning|error|emergency)_errno$";
+expression list args;
+expression e;
+local idexpression r;
+@@
+- log_LEVEL_errno(e, args);
+- r = e;
++ r = log_LEVEL_errno(e, args);
